core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -373,7 +373,7 @@ class SparkContext(config: SparkConf) extends Logging {
 
   private[spark] def cleaner: Option[ContextCleaner] = _cleaner
 
-  private[spark] var checkpointDir: Option[String] = None
+  private[spark] var checkpointDir: Option[String] = conf.getOption(CHECKPOINT_DIR.key)
```

Comment:
Initializing the variable is not sufficient.
Something like this perhaps:

```suggestion
  private[spark] var checkpointDir: Option[String] = None
  conf.getOption(CHECKPOINT_DIR).foreach(setCheckpointDir)
```